### PR TITLE
Make stream capturing test class opportunistic

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

In #5071 we tried to fix an implicit testing requirement dependency we
added in #3982 for capturing output streams from test runs. However,
that change didn't go far enough in making the capture features opt-in.
While it fixed the hard failure it still made the installation of
testtools and fixtures hard requirement for running tests, even if the
stream capturing was never used. For an example of this see #5078. This
commit attempts to remedy this situation by making the stream capturing
opportunistically opt-in. It splits out the common functionality between
the old test class without stream capturing into a new common base
class. Then it adds the stream capturing base class on top of that. The
stream capturing class is only used if testtools and fixtures is
installed and QISKIT_CAPTURE_STREAMS is not set.

### Details and comments

### CODE REVIEW

**Review by 1ucian0 — APPROVED** [review]

File: qiskit/test/base.py [hunk 1]
```diff
@@ -82,17 +82,94 @@ def gather_details(source_dict, target_dict):
...
+class BaseQiskitTestCase(unittest.TestCase):
+    """Common extra functionality on top of unittest."""
+    @staticmethod
+    def _get_resource_path(filename, path=Path.TEST):
+        """Get the absolute path to a resource.
+
+        Args:
+            filename (string): filename or relative path to the resource.
+            path (Path): path used as relative to the filename.
+        Returns:
```
Comment (1ucian0) [inline qiskit/test/base.py:94]:
```suggestion

        Returns:
```

**Review by 1ucian0 — DISMISSED** [review]

File: qiskit/test/base.py [hunk 2]
```diff
@@ -450,3 +476,8 @@ def valid_comparison(value):
         return error_msg[:-2] + msg_suffix
     else:
         return ''
+
+if not HAS_FIXTURES and not os.environ.get('QISKIT_TEST_CAPTURE_STREAMS'):
```
Comment (1ucian0) [inline qiskit/test/base.py:480]:
```suggestion

if not HAS_FIXTURES and not os.environ.get('QISKIT_TEST_CAPTURE_STREAMS'):
```

**Comment by abhobe** [comment]
There are linting errors,
I ran them locally on pylint.

`************* Module base
base.py:13:0: E0012: Bad option value 'missing-type-doc' (bad-option-value)
base.py:38:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
base.py:39:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
base.py:108:4: R0913: Too many arguments (7/5) (too-many-arguments)
base.py:152:51: E0602: Undefined variable 'inspect' (undefined-variable)
base.py:153:12: E0602: Undefined variable 'setup_test_logging' (undefined-variable)
base.py:158:8: C0415: Import outside toplevel (qiskit.providers.basicaer) (import-outside-toplevel)
base.py:160:8: W0212: Access to a protected member _backends of a client class (protected-access)
base.py:160:29: W0212: Access to a protected member _verify_backends of a client class (protected-access)
base.py:163:0: R0902: Too many instance attributes (9/7) (too-many-instance-attributes)
base.py:174:8: E1003: Bad first argument 'BasicQiskitTestCase' given to super() (bad-super-call)
base.py:201:35: W0212: Access to a protected member _UnexpectedSuccess of a client class (protected-access)
base.py:258:24: W0212: Access to a protected member _details of a client class (protected-access)
base.py:373:8: C0415: Import outside toplevel (qiskit.providers.basicaer) (import-outside-toplevel)
base.py:375:8: W0212: Access to a protected member _backends of a client class (protected-access)
base.py:375:29: W0212: Access to a protected member _verify_backends of a client class (protected-access)
base.py:449:8: R1705: Unnecessary "else" after "return" (no-else-return)
base.py:475:4: R1705: Unnecessary "else" after "return" (no-else-return)

-----------------------------------
Your code has been rated at 7.89/10
`

**Comment by abhobe** [comment]
@1ucian0 Any Updates on the merge?

**Review by 1ucian0 — APPROVED** [review]

**Pull request merged**